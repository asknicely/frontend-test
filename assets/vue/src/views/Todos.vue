<template>
  <div class="col-md-4 col-md-offset-4">
    <h1>{{ title }}</h1>
    <table class="table table-striped todos-list">
      <tbody>
        <tr>
          <th>#</th>
          <th>Description</th>
          <th align="center">User</th>
          <th width="20"></th>
          <th width="54"></th>
        </tr>
        <template v-if="activeId">
          <transition name="fade">
            <TodoItem
              @update-todos="load()"
              v-on:update:activeId="activeId = $event"
              :isActive="true"
              :id="filterdTodos[0].id"
              :user_id="filterdTodos[0].user_id"
              :description="filterdTodos[0].description"
              :completed="filterdTodos[0].completed"
              :baseurl="baseurl"
            />
          </transition>
        </template>
        <template v-else>
          <TodoItem
            v-for="(todo, idx) in orderedtodos"
            @update-todos="load()"
            v-on:update:activeId="activeId = $event"
            :key="idx"
            :id="todo.id"
            :user_id="todo.user_id"
            :description="todo.description"
            :completed="todo.completed"
            :baseurl="baseurl"
          />
        </template>
      </tbody>
      <tfoot>
        <tr v-if="!activeId">
          <td colspan="4">
            <div class="todos-list__description">
              <input
                v-model="todoDescription"
                type="text"
                name="description"
                class="small-6 small-center todos-list-description__input"
                placeholder="Description..."
              />
              <span class="focus-border"></span>
            </div>
          </td>
          <td>
            <button
              v-if="todoDescription"
              @click="add()"
              class="btn btn-sm todos-list__add"
            >
              Add
            </button>
            <button
              v-else
              type="submit"
              disabled
              class="btn btn-sm btn-primary disabled"
            >
              Add
            </button>
          </td>
        </tr>
        <tr v-else>
          <td colspan="5">
            <button @click="activeId = null" class="btn btn-sm btn-primary">
              Back
            </button>
          </td>
        </tr>
      </tfoot>
    </table>
  </div>
</template>
<script>
import orderBy from "lodash";
import TodoMixin from "../lib/Misc/TodoMixin";
import TodoItem from "../components/TodoItem";

export default {
  name: "Todos",
  mixins: [TodoMixin],
  components: { TodoItem },
  computed: {
    orderedtodos() {
      /*
        reorder todo list if necessary
      */
      return orderBy.orderBy(this.todos, "id");
    },
    filterdTodos() {
      /*
        show active todo
      */
      const result = this.todos.filter(obj => {
        return obj.id === this.activeId;
      });
      return result;
    },
    title() {
      /*
        page title
      */
      if (this.activeId) {
        return "Todo:";
      } else {
        return "Todo List:";
      }
    }
  },
  data() {
    return {
      baseurl: ""
    };
  },
  created() {
    this.baseurl = this.$attrs.baseurl;
    this.load();
  }
};
</script>
<style type="text/css">
.btn-primary {
  background-color: #3399ff;
  border-color: #3399ff;
}
.todos-list__add {
  background-color: #3399ff;
  color: #fff;
}
.todos-list__description {
  display: inline-block;
  position: relative;
}
.todos-list-description__input {
  border: 0;
  padding: 0.5rem 0rem;
  border: 0;
  border-bottom: 1px solid #ccc;
}
.todos-list-description__input:focus {
  outline: 0;
}
.todos-list-description__input ~ .focus-border {
  position: absolute;
  bottom: 0;
  left: 0;
  width: 0;
  height: 2px;
  background-color: #3399ff;
  transition: 0.4s;
}
.todos-list-description__input:focus ~ .focus-border {
  width: 100%;
  transition: 0.4s;
}
tfoot td {
  position: relative;
}
</style>
