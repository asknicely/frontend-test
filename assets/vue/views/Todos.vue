<template>
  <div class="col-md-4 col-md-offset-4">
    <h1>{{ title }}</h1>
    <table class="table table-striped">
      <tbody>
        <tr>
          <th width="54"></th>
          <th>#</th>
          <th>Description</th>
          <th>User</th>
          <th></th>
        </tr>
        <TodoItem
          v-for="(todo, idx) in orderedtodos"
          @update-todos="load()"
          :key="idx"
          :id="todo.id"
          :user_id="todo.user_id"
          :description="todo.description"
          :completed="todo.completed"
          :baseurl="baseurl"
        />
        <tr>
          <td colspan="4">
            <input
              v-model="todoDescription"
              type="text"
              name="description"
              class="small-6 small-center"
              placeholder="Description..."
            />
          </td>
          <td>
            <button
              v-if="todoDescription"
              @click="add()"
              class="btn btn-sm btn-primary"
            >
              Add
            </button>
            <button
              v-else
              type="submit"
              disabled
              class="btn btn-sm btn-primary disabled"
            >
              Add
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</template>
<script>
import lodash from "lodash";
import TodoMixin from "../lib/Misc/TodoMixin";
import TodoItem from "../components/TodoItem";

export default {
  name: "Todos",
  mixins: [TodoMixin],
  components: { TodoItem },
  computed: {
    orderedtodos() {
      return _.orderBy(this.todos, "id");
    }
  },
  data() {
    return {
      title: "Todo List:",
      baseurl: ""
    };
  },
  watch: {
    todos: {
      handler(val) {
        this.todos = val;
      },
      immediate: true
    }
  },
  created() {
    this.baseurl = this.$attrs.baseurl;
    this.load();
  }
};
</script>
<style type="text/css">
input[type=text] {
  padding: .5rem 1rem;
}
</style>
