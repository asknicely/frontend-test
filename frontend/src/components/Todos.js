import React, { useContext, useEffect, useState } from 'react'
import axios from 'axios'
import { toast } from 'react-toastify'
import { Context } from '../redux/app'
import Capitalize from '../functions/Capitalize'

import Loader from './Loader'
import Empty from './Empty'
import TodosItem from './TodosItem'
import TodosForm from './TodosForm'
import TodosHeader from './TodosHeader'

const Todos = (props) => {

	const [ todos, setTodos ] = useState({ list: [], isLoaded: false })
	const [ description, setDescription ] = useState('')
	const { state } = useContext(Context)
	// const todoId = props.match.params.id

	const fetchData = async (id) => {

		const url = `${process.env.REACT_APP_BACKEND_URL}/todo`
		const options = { headers: { 'Authorization': state.token }}
		const response = await axios.post(url, {}, options)

		if (response.data.statusTxt === 'error') {
			toast.error(Capitalize(response.data.msg))
			props.history.push('/login')
		}
		else {

			setTodos({
				list: response.data.result,
				isLoaded: true
			})

		}
	}

	const handleDelete = async (id) => {

		var formData = new FormData()
		formData.append('id', id)

		const url = `${process.env.REACT_APP_BACKEND_URL}/todo/delete`
		const options = { headers: { 'Authorization': state.token }}
		const response = await axios.post(url, formData, options)

		if (response.data.statusTxt === 'error') {
			toast.error(Capitalize(response.data.msg))
			props.history.push('/login')
		}
		else { 
			fetchData() 
			toast.success('Todo item deleted')
		}
	}

	const handleComplete = async (id) => {

		var formData = new FormData()
		formData.append('id', id)

		const url = `${process.env.REACT_APP_BACKEND_URL}/todo/complete`
		const options = { headers: { 'Authorization': state.token }}
		const response = await axios.post(url, formData, options)

		if (response.data.statusTxt === 'error') {
			toast.error(Capitalize(response.data.msg))
			props.history.push('/login')
		}
		else { 
			fetchData() 
			toast.success('Todo item marked as completed')
		}
	}
	
	const handleSubmit = async (e) => {

		e.preventDefault()
		if (description) {

			var formData = new FormData()
			formData.append('description', description)

			const url = `${process.env.REACT_APP_BACKEND_URL}/todo/add`
			const options = { headers: { 'Authorization': state.token }}
			const response = await axios.post(url, formData, options)

			if (response.data.statusTxt === 'error') {
				toast.error(Capitalize(response.data.msg))
				props.history.push('/login')
			}
			else { 

				setDescription('')
				fetchData() 
				toast.success('Todo item added')

			}		
		}
		else {
			toast.error('Please enter description to add a TODO')
		}
	}

	useEffect(() => { 

		if (!state.isLoggedIn) { props.history.push('/login') }
		else {

			if (!todos.isLoaded) {
				fetchData()
			}

		}

	})

	if (!todos.isLoaded) return (<Loader />)
	else {

		return (
			<div className="col-md-4 col-md-offset-4">
				<h1>Todo List:</h1>

				<form method="post" onSubmit={handleSubmit}>
				<table className="table table-striped">
					<tbody>
						
						<TodosHeader />

						{todos.list.map((todo) => {
							return (
								<TodosItem 
									key={todo.id}
									id={todo.id}
									user={todo.user_id}
									description={todo.description}
									completed={todo.completed}
									handleDelete={handleDelete}
									handleComplete={handleComplete}
									link={true}
								/>
							)
						})}

						{todos.list.length <= 0 ? <Empty /> : <React.Fragment></React.Fragment>}
		
						<TodosForm 
							description={description}
							setDescription={setDescription}
							handleSubmit={handleSubmit}
						/>

					</tbody>
				</table>
				</form>
			</div>
		)
	
	}
}

export default Todos