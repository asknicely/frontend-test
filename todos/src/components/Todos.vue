<template>
    <div class="login-page">
        <div class="form">
            <form class="login-form">
                <input type="text" placeholder="username"/>
                <input type="password" placeholder="password"/>
                fadsfafasf
                
                <button>login</button>
            </form>
        </div>
    </div>
</template>

<script>
    export default {
        name: 'Todos',
    }
</script>

