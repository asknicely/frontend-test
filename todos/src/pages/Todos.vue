<template>
    <div>
        <!--Navbar -->
           <Nav-Component></Nav-Component>
        <!--/.Navbar -->
        <b-alert
                :show="dismissCountDown"
                dismissible
                variant="success"
                @dismissed="dismissCountDown=0"
                @dismiss-count-down="countDownChanged"
        >
           {{todoTitle}} is removed
        </b-alert>
        <div class="login-page todo">
            <div class="form">
                <table class="table table-striped">
                    <th>#</th>
                    <th>User</th>
                    <th>Description</th>
                    <th>Completed</th>
                    <th></th>
                    
                    <Todo-Component v-for="(todo, index) in getTodos" v-bind:todo="todo" @onDeleted='showAlert(todo)'>
                      
                    </Todo-Component>
                    <TodoAdd-Component></TodoAdd-Component>
                 
                </table>
                
            </div>
        </div>
    </div>
    
</template>

<script>
    import Nav from '../components/Nav.vue'
    import Todo from '../components/Todo.vue'
    import TodoAdd from '../components/TodoAdd.vue'
    import { mapGetters } from "vuex";
    export default {
        name: 'Todos',
        computed: {
            ...mapGetters("user", ["getLoginedUser"]),
            ...mapGetters("todos", ["getTodos"])
        },
        data(){
            return {
                dismissCountDown: 0,
            }
        },
        mounted(){
            if(!this.getLoginedUser.id&&!this.getLoginedUser.username){
                this.$router.push('/');
            }
            this.$store.dispatch('todos/getTodosAsync');

        },
        methods:{
            countDownChanged(dismissCountDown) {
                this.dismissCountDown = dismissCountDown;
                if(dismissCountDown===0) {
                    this.todoTitle = null;
                }
            },
            showAlert(todo){
                this.dismissCountDown=5;
                this.todoTitle=todo.description; 
            }
        },
        components: {
            'Nav-Component': Nav,
            'Todo-Component':Todo,
            'TodoAdd-Component':TodoAdd,
            
        }
    }
</script>

