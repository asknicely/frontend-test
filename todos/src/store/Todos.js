import Vue from "vue";
import TodosService from '../services/todos'
const todosService = new TodosService('localhost:1337');

export default {
    namespaced: true,
    state: {
        todos:{}
    },

    getters:{
        getTodos(state){
            return state.todos
        },
    },
    
    mutations: {
        saveTodos(state, { todos }) {
            todos.forEach(i => {
                Vue.set(state.todos, i.id, i);
            });
        },
        deleteTodo(state, { todo }) {
            Vue.delete(state.todos, todo.id);
        },
    },
    actions: {
        async getTodosAsync({commit}){
            const response = await todosService.getTodos();
            const todos = response.data.map(todo=>{
                return {
                    ...todo,
                    completed:todo.completed==1? true: false,
                }
            });
            commit('saveTodos', {todos});
            return todos;
        },

        async completeTodoAsync({commit},{todo}){
            const response = await todosService.completeTodo(todo.id);
            return response;
        },

        async deleteTodoAsync({commit},{todo}){
            const response = await todosService.deleteTodo(todo.id);
            commit('deleteTodo', { todo });
            return response;
        },

        async addTodoAsync({commit},{description,userId}){
            const response = await todosService.addTodo({description,userId});
            
            return response;
        },
        
        
        
    }
}