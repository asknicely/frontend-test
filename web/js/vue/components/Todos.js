export default {
  data () {
    return {
      title: 'Todo List:',
      todos: [],
      baseurl: '',
      todoDescription: ''
    }
  },
  methods: {
    // Updates the dom by refreshing the contents of cart_items retrieved from the ajax api
    async load () {
      axios({
          method: 'GET',
          headers: { 'Content-Type': 'application/json' },
          url: '/todo',
          data: {}
        })
        .then(response => {
          if (response.status === 200) {
            this.todos = response.data
          } else {
            console.warn(response)
          }
        })
    },
    add () {
      axios({
          method: 'POST',
          url: `${this.baseurl}/todo/add`,
          headers: { 'Content-Type': 'application/json' },
          data: {
            'description': `${this.todoDescription}`
          }
        })
        .then(response => {
          if (response.status === 200) {
            this.$bus.$emit('updateTodoEvent')
          } else {
            console.warn(response)
          }
        })
    },
    complete (id) {
      axios
        .post(`${this.baseurl}/todo/complete/${id}`)
        .then(response => {
          if (response.status === 200) {
            this.$bus.$emit('updateTodoEvent')
          } else {
            console.warn(response)
          }
        })
    },
    remove (id) {
      axios
        .post(`${this.baseurl}/todo/delete/${id}`)
        .then(response => {
          if (response.status === 200) {
            this.$bus.$emit('updateTodoEvent')
          } else {
            console.warn(response)
          }
        })
    }
  },
  async created () {
    this.load()
    this.baseurl = this.$attrs.baseurl
  },
  mounted () {
    this.$bus.$on('updateTodoEvent', () => {
      this.load()
    })
  }
}
