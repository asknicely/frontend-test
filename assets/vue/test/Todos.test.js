import { mount } from "@vue/test-utils";
import Todos from "../src/views/Todos";

describe("Todos", () => {
  const wrapper = mount(Todos);

  it("renders the correct markup", () => {
    expect(wrapper.html()).toContain(
      '<table class="table table-striped todos-list">'
    );
  });

  it("has a button", () => {
    expect(wrapper.contains("button")).toBe(true);
  });
});
