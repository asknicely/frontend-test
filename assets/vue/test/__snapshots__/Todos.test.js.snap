// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todos.test.js has the expected html structure 1`] = `
<div
  class="col-md-4 col-md-offset-4"
>
  <h1>
    Todo List:
  </h1>
   
  <table
    class="table table-striped todos-list"
  >
    <tbody>
      <tr>
        <th>
          #
        </th>
         
        <th>
          Description
        </th>
         
        <th
          align="center"
        >
          User
        </th>
         
        <th
          width="20"
        />
         
        <th
          width="54"
        />
      </tr>
       
    </tbody>
     
    <tfoot>
      <tr>
        <td
          colspan="4"
        >
          <div
            class="todos-list__description"
          >
            <input
              class="small-6 small-center todos-list-description__input"
              name="description"
              placeholder="Description..."
              type="text"
            />
             
            <span
              class="focus-border"
            />
          </div>
        </td>
         
        <td>
          <button
            class="btn btn-sm btn-primary disabled"
            disabled="disabled"
            type="submit"
          >
            
            Add
          
          </button>
        </td>
      </tr>
    </tfoot>
  </table>
</div>
`;
